### PR TITLE
[Feature] Generic Radial Selector Menu

### PR DESCRIPTION
**Pull request by Remuchi**
# Description

Introduced a generic radial selector menu. Switched shortconstructionsystem from it's own BUI to radial selector.

# Changelog

no cl no fun (no changelog for players)

### CODE REVIEW

File: Content.Client/ShortConstruction/ShortConstructionSystem.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ShortConstructionComponent, RadialSelectorSelectedMessage>(OnItemRecieved);
+    }
+
+    private void OnItemRecieved(Entity<ShortConstructionComponent> ent, ref RadialSelectorSelectedMessage args)
+    {
+        if (!_proto.TryIndex(args.SelectedItem, out ConstructionPrototype? prototype))
+            return;
```
Comment (VMSolidus) [inline Content.Client/ShortConstruction/ShortConstructionSystem.cs:28]:
```suggestion
        if (!_proto.TryIndex(args.SelectedItem, out ConstructionPrototype? prototype)
            || !_gameTiming.IsFirstTimePredicted)
            return;
```

File: Content.Client/ShortConstruction/ShortConstructionSystem.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+using Robust.Shared.Enums;
+using Robust.Shared.Prototypes;
+
+namespace Content.Client.ShortConstruction;
+
+public sealed class ShortConstructionSystem : EntitySystem
+{
+    [Dependency] private readonly IPlacementManager _placement = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+
```
Comment (VMSolidus) [inline Content.Client/ShortConstruction/ShortConstructionSystem.cs:15]:
```suggestion
    [Dependency] private readonly IGameTiming _gameTiming = default!;
```

File: Content.Client/ShortConstruction/ShortConstructionSystem.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿using Content.Client.Construction;
+using Content.Shared.Construction.Prototypes;
+using Content.Shared.RadialSelector;
+using Content.Shared.ShortConstruction;
+using Robust.Client.Placement;
+using Robust.Shared.Enums;
+using Robust.Shared.Prototypes;
```
Comment (VMSolidus) [inline Content.Client/ShortConstruction/ShortConstructionSystem.cs:7]:
```suggestion
using Robust.Shared.Prototypes;
using Robust.Shared.Timing;
```

File: Content.Client/RadialSelector/RadialSelectorMenuBUI.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+using Content.Client.UserInterface.Controls;
+using Content.Shared.Construction.Prototypes;
+using Content.Shared.RadialSelector;
+using JetBrains.Annotations;
+using Robust.Client.GameObjects;
+using Robust.Client.Graphics;
+using Robust.Client.Input;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.Controls;
+using Robust.Shared.Prototypes;
```
Comment (VMSolidus) [inline Content.Client/RadialSelector/RadialSelectorMenuBUI.cs:11]:
```suggestion
using Robust.Shared.Prototypes;
using Robust.Shared.Timing;
```

File: Content.Client/RadialSelector/RadialSelectorMenuBUI.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            _menu.OpenCentered();
+        else
+            _menu.OpenCenteredAt(_inputManager.MouseScreenPosition.Position / _displayManager.ScreenSize);
+    }
+
+    protected override void UpdateState(BoundUserInterfaceState state)
+    {
+        base.UpdateState(state);
+
+        if (state is not RadialSelectorState radialSelectorState)
```
Comment (VMSolidus) [inline Content.Client/RadialSelector/RadialSelectorMenuBUI.cs:60]:
```suggestion
        if (state is not RadialSelectorState radialSelectorState
            || !_gameTiming.IsFirstTimePredicted)
```

File: Content.Client/RadialSelector/RadialSelectorMenuBUI.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+namespace Content.Client.RadialSelector;
+
+[UsedImplicitly]
+public sealed class RadialSelectorMenuBUI : BoundUserInterface
+{
+    [Dependency] private readonly IClyde _displayManager = default!;
+    [Dependency] private readonly IInputManager _inputManager = default!;
+    [Dependency] private readonly EntityManager _entManager = default!;
+    [Dependency] private readonly IPrototypeManager _protoManager = default!;
```
Comment (VMSolidus) [inline Content.Client/RadialSelector/RadialSelectorMenuBUI.cs:23]:
```suggestion
    [Dependency] private readonly IPrototypeManager _protoManager = default!;
    [Dependency] private readonly IGameTiming _gameTiming = default!;
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]
This is needed so that clicking on an item craft doesn't **instantaneously** craft between 7 and 14 of that item depending on your ping, and instead only crafts 1 per click. This also fixes the issue of the menu briefly following the cursor when opening it. The reason that was happening was because it was actually opening the menu 7 to 14 times in rapid succession under the mouse cursor due to misprediction.

Other than that, my testing shows that this system works just fine.

**Review by VMSolidus — APPROVED** [review]

**Review by Aidenkrz — APPROVED** [review]
need this for muh pr, maints merge faster, thanks!

**Review by Peptide90 — APPROVED** [review]

**Pull request merged**